crates/ark/src/plots/graphics_device.rs

```diff
@@ -458,96 +603,142 @@ macro_rules! with_device {
+/// copying the graphics display list to a new plot device, and then closing that device.
+#[tracing::instrument(level = "trace", skip_all)]
+unsafe extern "C-unwind" fn callback_deactivate(dev: pDevDesc) {
+    log::trace!("Entering callback_deactivate");
 
     DEVICE_CONTEXT.with_borrow(|cell| {
-        if let Some(callback) = cell._callbacks.deactivate.get() {
+        // We run our hook first to record before we deactivate the underlying device,
+        // in case device deactivation messes with the display list
+        cell.hook_deactivate();
+        if let Some(callback) = cell.wrapped_callbacks.deactivate.get() {
             callback(dev);
         }
     });
 }
 
-unsafe extern "C-unwind" fn gd_hold_flush(dev: pDevDesc, mut holdflush: i32) -> i32 {
-    log::trace!("gd_hold_flush");
+#[tracing::instrument(level = "trace", skip_all, fields(holdflush = %holdflush))]
+unsafe extern "C-unwind" fn callback_holdflush(dev: pDevDesc, mut holdflush: i32) -> i32 {
+    log::trace!("Entering callback_holdflush");
 
     DEVICE_CONTEXT.with_borrow(|cell| {
-        if let Some(callback) = cell._callbacks.holdflush.get() {
+        if let Some(callback) = cell.wrapped_callbacks.holdflush.get() {
             holdflush = callback(dev, holdflush);
         }
-
-        cell.holdflush(holdflush);
+        cell.hook_holdflush(holdflush);
         holdflush
     })
 }
 
 // mode = 0, graphics off
 // mode = 1, graphics on
 // mode = 2, graphical input on (ignored by most drivers)
-unsafe extern "C-unwind" fn gd_mode(mode: i32, dev: pDevDesc) {
-    log::trace!("gd_mode: {mode}");
+#[tracing::instrument(level = "trace", skip_all)]
+unsafe extern "C-unwind" fn callback_mode(mode: i32, dev: pDevDesc) {
+    log::trace!("Entering callback_mode");
 
     DEVICE_CONTEXT.with_borrow(|cell| {
-        if let Some(callback) = cell._callbacks.mode.get() {
+        if let Some(callback) = cell.wrapped_callbacks.mode.get() {
             callback(mode, dev);
         }
-        cell.mode(mode, dev);
+        cell.hook_mode(mode);
     });
 }
 
-unsafe extern "C-unwind" fn gd_new_page(dd: pGEcontext, dev: pDevDesc) {
-    log::trace!("gd_new_page");
+#[tracing::instrument(level = "trace", skip_all)]
+unsafe extern "C-unwind" fn callback_new_page(dd: pGEcontext, dev: pDevDesc) {
+    log::trace!("Entering callback_new_page");
 
     DEVICE_CONTEXT.with_borrow(|cell| {
-        if let Some(callback) = cell._callbacks.newPage.get() {
+        if let Some(callback) = cell.wrapped_callbacks.newPage.get() {
             callback(dd, dev);
         }
-        cell.new_page(dd, dev);
+        cell.hook_new_page();
     });
 }
 
 unsafe fn ps_graphics_device_impl() -> anyhow::Result<SEXP> {
     // TODO: Don't allow creation of more than one graphics device.
     // TODO: Allow customization of the graphics device here?
 
-    // TODO: Infer appropriate resolution based on whether display is high DPI.
-    let res = 144;
```

Comment:
Now set to `defaultResolutionInPixelsPerInch()` on the R side, which is 96 or 72 depending on the OS. I'm still not sure if that is exactly right but it fixes a bunch of "figure margins too big" issues.

Reply:
also, the high dpi inference should come from the pixel_ratio value